### ISSUE TITLE
calling `url` on a `Hash` (NoMethodError)

### ISSUE THREAD

**Issue by Cohen-Carlisle** [lostisland/faraday#628]
faraday-0.10.0/lib/faraday/connection.rb:60 can now call url on a hash because of the changes on the line above it.

```
     Failure/Error: response = Faraday.new(url: gateway_url).run_request(*request_parameters)

     NoMethodError:
       undefined method `url' for #<Hash:0x007f757ad12f58>
     # ./vendor/bundle/ruby/gems/faraday-0.10.0/lib/faraday/connection.rb:63:in `initialize'
     # ./vendor/bundle/ruby/gems/faraday-0.10.0/lib/faraday.rb:74:in `new'
     # ./vendor/bundle/ruby/gems/faraday-0.10.0/lib/faraday.rb:74:in `new'
     # ./app/middleware/gateway_proxy.rb:24:in `proxy_response'
```

0.9.2:
```
        options = ConnectionOptions.from(url)
        url     = options.url
```

0.10.0:
```
        options = options ? options.merge(url) : ConnectionOptions.from(url)
        url     = options.url
```

Called with code:
`response = Faraday.new(url: gateway_url).run_request(*request_parameters)`

**Comment by Cohen-Carlisle** [comment]
It looks like this stems from this setting:
`Faraday.default_connection_options = {}`
being a hash.

**Comment by Cohen-Carlisle** [comment]
Still digging... looks like we should be setting these as
```
Faraday.default_connection_options.request.open_timeout = 3
Faraday.default_connection_options.request.timeout = 10
```
instead of how we were...
```
Faraday.default_connection_options = { request: { open_timeout: 3, timeout: 10 } }
```

**Comment by iMacTia** [comment]
Hi @Cohen-Carlisle, I'm happy you were able to find out the issue on your own.
As you probably noticed already, `Faraday.default_connection_options` is an object of type `ConnectionOptions` and it's initialised like [this](https://github.com/lostisland/faraday/blob/974e0e122c223d4e74fe0fc7048ed41c12ef6ebd/lib/faraday.rb#L121)

```ruby
# Returns a Faraday::ConnectionOptions.
def self.default_connection_options
  @default_connection_options ||= ConnectionOptions.new
end
```

So when you re-assigned it to an hash like this:

```ruby
Faraday.default_connection_options = { request: { open_timeout: 3, timeout: 10 } }
```

you're loosing the ability to use the dot-notation to access his fields.
As you already understood, you simply need to set the options you need rather than using the `=` assign operator.

**Comment by Cohen-Carlisle** [comment]
@iMacTia after further investigation I think this actually is a bug introduced by this fix in `0.10.0`:
> Allow `default_connection_options` to be merged when options are passed as url parameter

In `0.9.2` using a hash to set `default_connection_options` was respected by the below syntax:
```
Faraday.default_connection_options = { request: { timeout: 0.0001 } }
Faraday.new("http://www.example.com/").get("/") # => Faraday::ConnectionFailed: execution expired
```
While it was not respected when using options passed as url parameters, but did not cause an exception:
```
Faraday.default_connection_options = { request: { timeout: 0.0001 } }
Faraday.new(url: "http://www.example.com/").get("/") # No exception, 200
```

Now the url params case tries to use `default_connection_options`, but does so without handling the possibility that the are given as a `Hash` instead of a `ConnectionOptions` object.
```
Faraday.default_connection_options = { request: { timeout: 0.0001 } }
Faraday.new(url: "http://www.example.com/").get("/") # NoMethodError: undefined method `url' for {:request=>{:timeout=>0.0001}, :url=>"http://www.example.com/"}:Hash
```
